File: leetcode/496. Next Greater Element I/solution3.py
class Solution(object):
    def nextGreaterElement(self, findNums, nums):
        """
        :type findNums: List[int]
        :type nums: List[int]
        :rtype: List[int]
        """
        next_greater_map = {}
        for i, n in enumerate(nums):
            for j in xrange(i+1, len(nums)):
                if nums[j] > n:
                    next_greater_map[n] = nums[j]
                    break
            else:
                next_greater_map[n] = -1
        return map(next_greater_map.get, findNums)

# 17 / 17 test cases passed.
# Status: Accepted
# Runtime: 49 ms
# 97.53%

File: algorithms/552. Student Attendance Record II/solution3.cc
class Solution {
public:
    int checkRecord(int n) {
        // based on @lixx2100
        int m = 1000000007;
        int f[100001][2][3] = {1, 1, 1, 1, 1, 1};
        for ( int i = 1; i <= n; ++i ) {
            for ( int j = 0; j < 2; ++j ) {
                for ( int k = 0; k < 3; ++k ) {
                    // append P
                    f[i][j][k] = f[i - 1][j][2];
                    if ( j > 0 ) {
                        // append A
                        f[i][j][k] = (f[i][j][k] + f[i - 1][j - 1][2]) % m;
                    }
                    if ( k > 0 ) {
                        // append L
                        f[i][j][k] = (f[i][j][k] + f[i - 1][j][k - 1]) % m;
                    }
                }
            }
        }
        return f[n][1][2];
    }
};

// 58 / 58 test cases passed.
// Status: Accepted
// Runtime: 149 ms
